src/models/FunctionTools.h

```diff
@@ -0,0 +1,59 @@
+//
+// Created by prova on 26.04.21.
+//
+
+#pragma once
+
+#include "PTRef.h"
+#include "SSort.h"
+class FunctionSignature {
+    friend class TemplateFunction;
+    SRef ret_sort;
+    std::string name;
+    vec<PTRef> args;
+private:
+    FunctionSignature() : ret_sort(SRef_Undef) {}
+public:
+    FunctionSignature(std::string && name, vec<PTRef> && args, SRef ret_sort)
+    : ret_sort(ret_sort)
+    , name(std::move(name))
+    , args(std::move(args))
+    {}
+    std::string getName() const { return name; }
+    SRef getRetSort() const { return ret_sort; }
+    const vec<PTRef>& getArgs() const { return args; }
+};
+
+class TemplateFunction {
+    FunctionSignature signature;
+    PTRef tr_body;
+public:
+    TemplateFunction(const std::string & name, const vec<PTRef>& args_, SRef ret_sort, PTRef tr_body)
+            : tr_body(tr_body)
+    {
+        args_.copyTo(signature.args);
+        signature.name = name;
+        signature.ret_sort = ret_sort;
+    }
+    TemplateFunction(FunctionSignature&& signature, PTRef body) : signature(std::move(signature)), tr_body(body) {}
+    TemplateFunction() : tr_body(PTRef_Undef) {}
+    TemplateFunction(const TemplateFunction& other) : tr_body(other.tr_body) {
+        signature.ret_sort = other.getRetSort();
+        signature.name = other.getName();
+        other.getArgs().copyTo(signature.args);
+    }
+    TemplateFunction& operator= (TemplateFunction&& other) noexcept {
```

Comment:
If it needs move assignment operator, should we add also move constructor?
Wouldn't actually the default move-assignment operator work here? It would require the move-assignment on FunctionSignature but that could also be default, no?

Reply:
I think that the default doesn't work for FunctionSignature because it has `mtl::vec` as a field.

The move assignment (or move constructor) is used from `Logic::defineFun`.  I'll add the mode constructor.

Reply:
But `mtl::vec` has both move constructor and move-assignment operator, so I still think it should work, no?